tokio/src/io/util/copy_bidirectional.rs

```diff
@@ -0,0 +1,116 @@
+cfg_io_util! {
+    use super::copy::CopyBuffer;
+
+    use crate::io::{AsyncRead, AsyncWrite};
+
+    use std::task::{Context, Poll};
+    use std::io;
+    use std::pin::Pin;
+    use std::future::Future;
+
+    enum TransferState {
+        Running(CopyBuffer),
+        ShuttingDown(u64),
+        Done(u64)
+    }
+
+    struct CopyBidirectional<'a, A: ?Sized, B: ?Sized> {
+        a: &'a mut A,
+        b: &'a mut B,
+        a_to_b: TransferState,
+        b_to_a: TransferState
+    }
+
+    fn transfer_one_direction<A, B>(
+        cx: &mut Context<'_>,
+        state: &mut TransferState,
+        r: &mut A,
+        w: &mut B
+    ) -> Poll<io::Result<u64>>
+        where A:  AsyncRead + AsyncWrite + Unpin + ?Sized, B: AsyncRead + AsyncWrite + Unpin + ?Sized
+    {
+        let mut r = Pin::new(r);
+        let mut w = Pin::new(w);
+
+        loop {
+            match state {
+                TransferState::Running(buf) => {
+                    let count = ready!(buf.poll_copy(cx, r.as_mut(), w.as_mut()))?;
+                    *state = TransferState::ShuttingDown(count);
+                }
+                TransferState::ShuttingDown(count) => {
+                    ready!(w.as_mut().poll_shutdown(cx))?;
+
+                    *state = TransferState::Done(*count);
+                }
+                TransferState::Done(count) => return Poll::Ready(Ok(*count)),
+            }
+        }
+    }
+
+    fn transpose<R, E>(r: Poll<Result<R,E>>) -> Result<Poll<R>, E> {
+        match r {
+            Poll::Pending => Ok(Poll::Pending),
+            Poll::Ready(Ok(r)) => Ok(Poll::Ready(r)),
+            Poll::Ready(Err(e)) => Err(e),
+        }
+    }
+
+    impl<'a, A, B> Future for CopyBidirectional<'a, A, B>
+        where A: AsyncRead + AsyncWrite + Unpin + ?Sized, B: AsyncRead + AsyncWrite + Unpin + ?Sized
+    {
+        type Output = io::Result<(u64, u64)>;
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            // Unpack self into mut refs to each field to avoid borrow check issues.
+            let CopyBidirectional { a, b, a_to_b, b_to_a } = &mut *self;
+
+            let a_to_b = transpose(transfer_one_direction(cx, a_to_b, &mut *a, &mut *b))?;
+            let b_to_a = transpose(transfer_one_direction(cx, b_to_a, &mut *b, &mut *a))?;
+
+            let a_to_b = ready!(a_to_b);
+            let b_to_a = ready!(b_to_a);
+
+            Ok((a_to_b, b_to_a)).into()
+        }
+    }
+
+    /// Copies data in both directions between `a` and `b`.
+    ///
+    /// This function returns a future which will read from both streams, and
+    /// write any data read to the opposing stream. The data transfer in one
+    /// direction (eg, `a` to `b`) will not be blocked if the opposite direction
+    /// blocks (e.g. if writing from `b` to `a` is blocked, `a` to `b` will not
+    /// be blocked).
```

Comment:
I try to avoid using the word "blocks" because it is ambiguous in the context of async/await. If one of them blocks the thread (doesn't return from `poll_read` or `poll_write`), the other direction is definitely blocked by that.

Reply:
Maybe just simplify a bit to this?

> This function returns a future that will read from both streams, writing any data read to the opposing stream. This happens in both directions concurrently.